File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        if n == 1: return 0
        buy = [0 for i in range(n)]
        sell = [0 for i in range(n)]
        buy[0] = -prices[0]
        buy[1] = max(buy[0], -prices[1])
        sell[1] = max(sell[0], prices[1] + buy[0])
        for i in range(2, n):
            buy[i] = max(buy[i-1], sell[i-2] - prices[i])
            sell[i] = max(sell[i-1], prices[i] + buy[i-1])
        return max(buy[n-1], sell[n-1])

File: 0309-best-time-to-buy-and-sell-stock-with-cooldown/0309-best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        int buy[n];
        int sell[n];
        memset(buy, 0, sizeof(buy));
        memset(sell, 0, sizeof(sell));
        buy[0] = -prices[0];
        sell[0] = 0;
        if(n == 1) return sell[0];
        buy[1] = max(-prices[1], buy[0]);
        sell[1] = max(prices[1] + buy[0], 0);
        for(int i=2; i<n; i++) {
            buy[i] = max(buy[i-1], sell[i-2] - prices[i]);
            sell[i] = max(prices[i] + buy[i-1], sell[i-1]);
        }
        return max(buy[n-1], sell[n-1]);
    }
};